### PR TITLE
fixed delay updating parameters;

### PR DESCRIPTION
**Pull request by zetta83**

### CODE REVIEW

**Comment by MicheleBertoli** [comment]
Hello @zetta83, thanks for your PR.
Can you please provide more context on this?

**Comment by zetta83** [comment]
for the setState callback you should actually use `this.props`
since it’s called after the `setState` is performed, the props from params may be old

**Comment by MicheleBertoli** [comment]
@zetta83 was this causing any real problem?
I don't see any open issue around this.
Is it possibile for you to provide a failing test or a minimal repro example?
Thanks!

**Comment by zetta83** [comment]
The `start` function is being called on `componentWillReceiveProps`, which means `this.props` relates to the previous component props, not the next props. If you set the `from` prop to `5` and then set it to `10`, the start value of the counter will be set to `5`, not `10`.
That means that the counter will always go back a little if we update the `from` prop.
Please check this pen showing the bug: https://codesandbox.io/s/PNjwWL5n2. This pen will run the counter from 0 to 10, then from 10 to 50, and then from 50 to 100.

**Comment by MicheleBertoli** [comment]
I see, thanks!
Shouldn't be enough to change the lifecycle method from `componentWillUpdate` to `componentDidUpdate`?

**Comment by zetta83** [comment]
that would work too, but it will make the component re-render every time it updates :slightly_smiling_face:
calling setState on `componentWillReceiveProps` won’t trigger another render

**Comment by MicheleBertoli** [comment]
Cool @zetta83, can you please replace the current "when receive new props" test block with this before I merge?

```js
describe('when receive new props', () => {
  class Parent extends Component {
    constructor() {
      super();
      this.state = {
        from: 0,
        to: 1,
      };
    }
    render() {
      return <CountTo speed={1} {...this.state} />;
    }
  }

  it('starts from 0, restarts from 0', () => {
    const parent = TestUtils.renderIntoDocument(
      <Parent />
    );
    const span = TestUtils.findRenderedDOMComponentWithTag(parent, 'span');
    expect(findDOMNode(span).textContent).toEqual('0');
    jest.runAllTimers();
    expect(findDOMNode(span).textContent).toEqual('1');
    parent.setState({
      to: 2,
    });
    expect(findDOMNode(span).textContent).toEqual('0');
    jest.runAllTimers();
    expect(findDOMNode(span).textContent).toEqual('2');
  });

  it('starts from 0, restarts from 2', () => {
    const parent = TestUtils.renderIntoDocument(
      <Parent />
    );
    const span = TestUtils.findRenderedDOMComponentWithTag(parent, 'span');
    expect(findDOMNode(span).textContent).toEqual('0');
    jest.runAllTimers();
    expect(findDOMNode(span).textContent).toEqual('1');
    parent.setState({
      from: 2,
      to: 3,
    });
    expect(findDOMNode(span).textContent).toEqual('2');
    jest.runAllTimers();
    expect(findDOMNode(span).textContent).toEqual('3');
  });
});
```

The second fails on master and passes on your branch.
It's important to cover the issues with tests when they are fixed to make sure they don't happen again.

File: src/__tests__/react-count-to-test.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import { findDOMNode } from 'react-dom';
 import TestUtils from 'react-dom/test-utils';
 import CountTo from '../react-count-to';
 
+jest.useFakeTimers();
```
Comment (MicheleBertoli) [inline src/__tests__/react-count-to-test.js]: This is unnecessary in Jest 12

**Pull request merged**